enhancements/baremetal/baremetal.md

```diff
@@ -0,0 +1,279 @@
+
+## Proposal
+
+Every OpenShift based Baremetal IPI deployment will run a "metal3" pod on
+one Master Node. A "metal3" pod includes a container running BareMetal
+Operator(BMO) and several other supporting containers that work together.
+
+The BMO and other supporting containers together are able to discover a
+baremetal server in a pre-determined provisioning network, learn the
+HW attributes of the server and eventually boot it to make it available
+as a Machine within a MachineSet.
+
+The Machine API Operator (MAO) currently deploys the "metal3" pod only
+when the Platform type is `BareMetal` but the BaremetalHost CRD is exposed
+by the MAO as part of the release payload which is managed by the cluster
+version operator.
+
+The installer is responsible for kicking off a Baremetal IPI deployment
+with the right configuration.
+
+### User Stories
+
+With the addition of features described in this and other enhancements
+detailed in this current directory, OpenShift can be used to bring up
+a functioning cluster starting with a set of baremetal servers. As
+mentioned earlier, these enhancements rely on the Baremetal Operator (BMO)
+[1] running within the "metal3" pod to manage baremetal hosts. The BMO in
+turn relies on the Ironic service [3] to manage and provision baremetal
+servers.
+
+1. Will enable the user to deploy a control plane with 3 master nodes.
+2. Will enable the user to grow the cluster by dynamically adding worker
+nodes.
+3. Will enable the user to scale down the cluster by removing worker nodes.
+
+### Implementation Details/Notes/Constraints
+
+Baremetal IPI is integrated with OpenShift through the metal3.io [8] project.
+Metal3.io is a set of Kubernetes controllers that wrap the OpenStack Ironic
+project to provide Kubernetes native APIs for managing deployment and
+monitoring of physical hosts.
+
+The installer support for Baremetal IPI deployments is described in more detail
+in [7]. The installer runs on a special "provisioning host" that needs to be
+connected to both a "provisioning network" and an "external network". The
+provisioning network is a dedicated network used just for the purposes of
+configuring baremetal servers to be part of the cluster. The traffic on the
+provisioning network needs to be isolated from the traffic on the external
+network (hence 2 seperate networks.). The external network is used to carry
+cluster traffic which which includes cluster control plane traffic, application
+and data traffic.
+
+Control Plane Deployment
+
+1. A minimin Baremetal IPI deployment consists of 4 hosts, one to be used
+first as a provisioning host and later potentially re-purposed as a worker.
+The other 3 make up the control plane. These 4 hosts need to be connected
+to both the provisioning and external networks.
+
+2. Installation can be kicked off by downloading and running
+"openshift-baremetal-install". This image differs from the "openshift-install"
```

Comment:
This is a deficiency - how are you planning to fix this?

Reply:
The reason we're not part of the main openshift-install binary is our dependency on libvirt libraries. One proposal we discussed previously is to use https://github.com/digitalocean/go-libvirt which just hits the RPC directly, but we'd have to spend the time and rewrite the libvirt terraform provider.

This isn't on our radar to fix in the near term. There's a spike on the installer team to investigate removing the bootstrap host which would make this easier. 

Do you want this documented here?

Reply:
@smarterclayton, as mentioned by @stbenjam removing the bootstrap node would remove our dependency on libvirt. That way baremetal IPI installs can be part of the Openshift installer. This is part of our road map. I have updated the document to reflect that.

I think this "enhancement" is essentially provides a background that reviewers of baremetal IPI related code patches need to get context for what they are reviewing. We will make sure to keep this document alive. For these reasons, I think it would be helpful to merge this document.

Reply:
Where and how is this separate installer maintained?

Reply:
It is a separate installer binary, it is part of the same repo. 

It is distributed as part of the baremetal-installer image (like libvirt-installer is distributed as part of libvirt-installer). It's extractable using `oc` by extracting the `openshift-baremetal-install` command.

@sadasu You may want to include this information here.